rule/rule.go

```diff
@@ -396,9 +396,12 @@ func (d ruleData) toAuditRuleData() (*auditRuleData, error) {
 	}
 
 	if d.allSyscalls {
-		for i := 0; i < len(rule.Mask)-1; i++ {
+		for i := range rule.Mask {
```

Comment:
> for i := range &rule.Mask { to avoid the potential (minor) allocation

How does that work? What allocation does it avoid? I'm happy to make the change. I just want to understand the internals. In comparing the assembly output between the two code versions (using a standalone file with the same loop as a proxy) there was no difference. I tried with and without optimizations (gcflag `-N`).

Reply:
You are absolutely correct; that's a part of the spec that I missed a change in (it's not been like [that](https://github.com/golang/go/blob/82d1a9dce7478abcd9cf5d3fdc94f231f2f7b614/doc/go_spec.html#L3990-L3993) since 2011. The relevant part of the current spec is:

>The range expression `x` is evaluated once before beginning the loop, with one exception: if at most one iteration variable is present and `len(x)` is [constant](https://go.dev/ref/spec#Length_and_capacity), the range expression is not evaluated.

In the case that both the index and value variables are used, then the array needs to be copied to evaluate the expression, this is not happening here. Sorry for the noise.

Reply:
Thanks for explaining.